### DISCUSSION TITLE
UI Revamp: Design System for CKAN

### DISCUSSION THREAD

**Discussion by thegostev** [ckan/ckan#8188] [CKAN 3.0]
I happy to share next deliverable produced by UI Revamp team - the design system with components in Figma.

It consists of: colors defined, new font, variables and components. 

Design system is open for commenting and sharing your feedback. Click Figma logo at top left of the page to access a menu with pages included: https://www.figma.com/file/oI3wDSdZ0cHZLh6CrfYMnb/CKAN-3.0-Design-System?type=design&node-id=0%3A1&mode=design&t=72fVNsR2fnUgkdrV-1

This design system will be used to create 6 basic layouts which then will be implemented. Having all components created first, making creating layouts faster.

Design system is a living document. We expect 25-50% of components to be updated during the layouts design and implementation stage, so they look great together.

Please share your feedback, ideas that you have and important info that we should include into the work. You can comment both here and in Figma.

Design approach:

1. All deliverables (artifacts) should be open to public with ability to leave feedback right in the artifact.
2. Tight collaboration between a frontend and a designer is essential to deliver an efficient solution.
3. **We should not make big changes, rather an update that doesn't touch how the system works now.**
4. Higher level of customization (colors, fonts) is always used, it should be taken into account.
5. UI changes should not make it harder to customize CKAN (as big portion of maintainers don't use out-of-the-box UI).
6. Documentation (screenshots) should be updated after new design is rolled out.
7. Work on the team cohesion early on, frequently communicating with other contributors.
8. It's beneficial to have a professional in the team who has experience with accessibility standards and CKAN.

Design requirements:

1. Font should support maximum diversity in languages.
2. UI should be designed to adapt a layout to different screen sizes and resolutions.
3. Navigation should be predictable and consistent across the website. Menus should be structured logically and standard elements like search bars and contact information should be easy to locate.
4. Provide adequate color contrast between text and background colors for better readability. The WCAG recommends a contrast ratio of at least 4.5:1 for normal text and 3:1 for large text.
5. Text should be resizable up to 200% without loss of content or functionality. Fonts should be legible, and paragraphs should have adequate spacing.
6. Target size for pointer inputs are 24x24 px.
7. Interactive elements should be operable through keyboard interfaces.
8. Images (unless purely decorative) should have alternative text that describes the image’s purpose or content.
9. UI should work across [major browsers](https://gs.statcounter.com/browser-market-share) (Chrome, Safari, Edge, Firefox).
10. Provide feedback for user interactions, such as loading indicators for long processes.
11. Accessibility standards should be taken into account to the level that the work completes requirements of WCAG 2.2 AA at 50% minimum.

**Comment by thegostev** [comment]
Work on layouts is under way. Today we discussed first two of them. Indeed they were done fast. Though when you have a layout done, right away it's visible what can be updated. So the updates are following immediately.

The plan is to have first 2-3 layouts, with at least 2 versions of each and 3 color schemas to pick from. Colors were defined during the design system creation though now it's possible to evaluate them looking on an actual UI colored.

The layouts should be presented soon. We'd decide tomorrow before core dev team meeting.

**Comment by thegostev** [comment]
Layouts for homepage, datasets list and organizations list are done and ready for receiving the feedback 📢 Just click on comment button in header and leave it right in the Figma file: https://www.figma.com/file/oI3wDSdZ0cHZLh6CrfYMnb/CKAN-3.0-Design-System?type=design&node-id=508-40&mode=design&t=q15o0VspaTkYmM29-0

Next steps: more layouts and feedback processing.

**Comment by bjornhagstrom** [comment]
On design requirement 11 "Accessibility standards should be taken into account to the level that the work completes requirements of WCAG 2.2 AA at 50% minimum.". I do not quite understand. Are you only striving to achieve a 50% fullfillment of WCAG 2.2 AA? Why not 100% CKAN should support 100% compliance with WCAG but if the users change things that might be impacted of cause but the standard design should be pretty close to perfect right? 

I like what I see in figma right now. The resultpage needs som love though. The colors for the features on a search hit (pdf, json e.t.c.) is a bit to dominant. The search page as a whould would benefit from some love.

It would also be great if there is a template for an accessibilty statment that is valid for a core ckan implementation and if there is a way to add to that page depending on what extentions you have installed (they provide their own partial statements).

**Reply by thegostev** [reply]
Hey, @bjornhagstrom, short answer is that WCAG 2.2 AA compatibility wasn't planned in the scope of UI Revamp project. That's why it's formulated as 50%, with the idea that we'd do what we can, along the way, leaving work on compatibility to a separate project. 

This separate project to achieve WCAG 2.2 AA, I'm planning right now and ideally we're starting with it in 2 months period. There are 55 criteria to comply to so I expect it'd include combined effort of designers and developers. It would be great to get external audit at the end of the project, though realistically it'll be a self-audit which can result in producing and publishing a report.

Thanks for getting into the conversation! Please let me know if you have anything to add or to discuss.

**Comment by thegostev** [comment]
Design system is built, along with layouts, feedback is taken and processed. The last design touch would be in the area of mobile interfaces.

Meanwhile we're moving to implementing existing designs, which is 3rd stage of UI Revamp initiative of 3 stages in total.

**Comment by thegostev** [comment]
We’d completed mobile designs. They’re located at the left-hand side of each layout page. Please leave your feedback right there in Figma: https://www.figma.com/design/oI3wDSdZ0cHZLh6CrfYMnb/CKAN-3.0-Design-System?node-id=994-20627&t=vuaKK2MlalWkI2ol-0

**Comment by thegostev** [comment]
132 accessibility-related issues found by scanning layouts in our Figma files. Fixes in Figma planned by next Monday, minor updates in JS/CSS expected.

**Comment by thegostev** [comment]
70% of implementation work completed [according to the plan](https://docs.google.com/spreadsheets/d/1FwZwLLDJcXsu-bnlHVj7Ba1M7Ft4JIRK0IfkQC218Jw/edit?gid=0#gid=0). 

Latest updates:

- Issue with [Noto Sans font licensing](https://github.com/ckan/ckan/discussions/8486).
- [Dataset creation workflow design updates](https://www.figma.com/design/oI3wDSdZ0cHZLh6CrfYMnb/CKAN-3.0-Design-System?node-id=2449-12526&node-type=frame&t=nBgsDc9hnQSsTVUq-0) to include changes introduced in CKAN 2.11.
- Login implemented.

UI Revamp team collaborates with CKAN Core Dev Team on a weekly basis by visiting joint meetings (for 3 weeks in a row), creating pull requests (rather monthly), discussing design deliverables in Figma (no particular cadence). 

Collaboration part is immensely important as with UI Revamp we're piloting community-led projects.

**Reply by thegostev** [reply]
P.S. You can check the progress on the public CKAN dev instance: 

https://ckan-design-preview.galv1.links.com.au/ 
Login: ckan
Password: preview

**Reply by fpichardom** [reply]
The new UI is looking great. Is there a repo with CKAN that is currently this new UI in progress? Is the code used for this preview instance private for now? When you mentioned, "You can check the progress on the public CKAN dev instance," I thought it was already available in the branch `dev-2.11`, but I don't see it.

Keep the good work, cheers,

Fritz

**Reply by aleeexgreeen** [reply]
Hi @fpichardom, the code is incrementally being added to `master`, we have #8503 and #8492 to start getting the new UI into the code. In `master`, we have added new folders for the UI, `templates-midnight-blue` and `public-midnight-blue`. As more components get added, you can view on your own instance by replacing public and templates as the base config folders, as shown below 
```
ckan.base_public_folder = public-midnight-blue
ckan.base_templates_folder = templates-midnight-blue
```
The in progress code is being worked on in multiple repos but the branch that is on the preview site can be found [here](https://github.com/aleeexgreeen/ckan/tree/midnight_blue_scss)

Thanks & looking forward to hearing your feedback!

**Comment by thegostev** [comment]
As for now the code is actively integrated in to the ckan/ckan repository. Integration comes with changes and updates both to the frontend code and to design. 

80% of the scope is integrated with new PR's with updates are submitted at least each month for the last 3 months.

**Comment by thegostev** [comment]
A massive batch of changes is integrated into CKAN repository, it consists of 32 commits with a big amount of components and a dataset creation page by Stephen Oni: [#8503](https://github.com/ckan/ckan/pull/8503).

**Comment by ThrawnCA** [comment]
We have some concerns about the changes made in https://github.com/ckan/ckan/issues/7586

- The resource "Manage" button was renamed to "Edit resource" and moved from the right to the left. This makes it inconsistent with the dataset "Manage" button, and also misleading since the button is still taking you to _all_ the resource management tabs - Data Dictionary, Datastore, etc - of which the "Edit resource" tab is only one.
- The "Add new resource" button was made available on more pages, which is nice, but it was also made smaller, in a less contrasting theme, and on the opposite side of the page from where it was before, as well as the opposite side from the dataset "Manage" button that the user probably clicked immediately beforehand - all of which makes it difficult for new users to spot. Also, it now seems to be available _only_ from the sidebar, never in the primary content section of the page, despite being a central feature, which to my understanding is not how a sidebar should work. Perhaps it could be added to the dataset view page alongside the Manage button?

**Reply by wardi** [reply]
I asked @ThrawnCA to post this comment here as well to get feedback from the team working on the new UI.

In #7586 I arranged the buttons in the on the resource page so they appear in the same order as when editing on the old UI:
![Screenshot from 2025-06-12 13-05-38](https://github.com/user-attachments/assets/3a78ba42-40ea-44c2-96a0-9a6edf346180)
![Screenshot from 2025-06-12 13-06-07](https://github.com/user-attachments/assets/0c95c657-e572-4778-accc-1d11dcc252b4)

new UI:
![Screenshot from 2025-06-12 13-13-05](https://github.com/user-attachments/assets/8f13f55c-adf7-4231-8480-4610705a69f2)
![Screenshot from 2025-06-12 13-13-29](https://github.com/user-attachments/assets/24e49f1d-3285-448c-9dac-1b5dcafe7e21)

I think this makes sense, even if that means the edit button appears on the left when looking at a resource and the manage button appears on the right when looking at the dataset page.

For the second point I think we could add an "Add new resource" button on the dataset read page under the list of resources in the main content of the page, but for A11y on the pages with the sidebar we shouldn't have a duplicate "Add new resource" button that goes to the same place.

We could certainly make the button larger or more prominent on the sidebar than it is now.

**Reply by ThrawnCA** [reply]
What about putting an Add New Resource button, in a suitably prominent style, next to the dataset Manage button? Currently that page doesn't have one, and it would shorten the process for a very common task.

![image](https://github.com/user-attachments/assets/728ee205-d6dd-43d5-aa42-b6976ace8fcd)

**Reply by aleeexgreeen** [reply]
For the second point, I agree adding the add new resource on the dataset page would shorten the process. This would also give us the opportunity to match the button order/positioning of the resource page. The "add new resource" button could be on the right in the prominent colour and manage could be on the left. Consistent design across those two pages.

Also when it comes to a11y, I'm only aware of redundant links (adjacent links leading to the same page) and links with different text leading to the same page as a11y violations. I think having multiple links with the same text on the same page is okay as long as there are not next to each other, but if we do not want to have multiple links on the same page, that's completely understandable. (I may be wrong though)

**Reply by aleeexgreeen** [reply]
Is there a desire to add the "Add new resource" button to the resource edit/resource views pages, beside the "View resource" button?
![image](https://github.com/user-attachments/assets/1349db30-61f8-4a37-a990-77d1c2174bc1)
I've added the "no_new_res" condition on the resource list in the sidebar which makes it possible to remove the "Add new resource" button there and add it elsewhere. It may not be consistent with the resource page resource list though.

**Reply by ThrawnCA** [reply]
> Is there a desire to add the "Add new resource" button to the resource edit/resource views pages

My understanding is that since there is already a sidebar link, doubling up creates a problem on screen readers.

**Comment by wardi** [comment]
Would like the team's thoughts on a new suggested toast module #9016 and alert dialog #9018 . Do these fit with the new theme or should they be modified?

**Reply by mutantsan** [reply]
I might be biased, but I will tell my thoughts anyway 🙂

If we didn’t move from the bootstrap 5, I don’t see why it shouldn’t fit. I didn’t use any custom css, besides 1 animation keyframe. If @aleeexgreeen will restyle the modals and toasts from the bootstrap, it will be applied automatically.

**Reply by aleeexgreeen** [reply]
I have done a bit of work on the scss for toasts and pop up dialogs/modals and I made very little changes because the designers essentially used bootstraps scss for both. The changes should already be in effect, I will take a look but I think the PRs should fit with the theme.
